### PR TITLE
Re-structuring the hyparview manager (wip)

### PR DESCRIPTION
**Pull request by junghunyoo**

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
Can you summarize what this PR is doing?

**Comment by junghunyoo** [comment]
This is a bit experiment. I am not sure whether this is the right change, but I made the sending message function synchronous (make sure the tcp delivers messages). And, I added self healing functionality: if the active view contains less than (max active view size / 2), the node sends the neighbor_request message. From my experience, this is the most stable implementation. The only problem which this change cannot fix is the partition (some nodes (not a single node) can be isolated) - I saw this only once, though.

**Comment by cmeiklejohn** [comment]
What's left here?

**Comment by junghunyoo** [comment]
If the random promotion is working without continuous promotions, I think there is nothing to be done with this PR for the functionality point of view. However, I believe that the detailed code review is needed for this change.

**Comment by cmeiklejohn** [comment]
@junghunyoo Can you make it so the random promotion is configurable via partisan_config?  That way we can merge this and enable/disable as necessary?

**Comment by junghunyoo** [comment]
@cmeiklejohn I will make that change during the weekend and let you know once it is done.

**Comment by cmeiklejohn** [comment]
@junghunyoo Thanks.

**Comment by cmeiklejohn** [comment]
@junghunyoo thanks bud.
